### DISCUSSION TITLE
Have a template macro that uses XML?

### DISCUSSION THREAD

**Discussion by alshdavid** [Relm4/Relm4#661] [General]
I'd love it if we had a template macro syntax that could convert xml-like syntax into GTK widgets (like JSX for the browser). 

The tree style `view!` macro by relm4 is already pretty close to that but I would love to go one step further with something like:

```rust
fn build_ui(app: &gtk::Application) {
  // ... 

  let root = relm4::rsx!(
    <gtk::Label>Hello World</gtk::Label>
  );

  // ...
  window.set_child(&root);
}
```

### Off topic brain dump

I ask because I have the ambition to implement react-like syntax on top of it, adding reactive state management controls, for example:

```rust
fn my_component() -> gtk::Box {
  let counter = SignalCell::new(0); // Rc<RefCell<T>> with a ".subscribe()" to notify when a change it made

  let on_click = {
    let counter = counter.clone();
    move |_this: &gtk::Button| {
      counter.fetch_update(|current| {
        (*current) += 1;
      })
    }
  };

  relm4::rsx!(
    <gtk::Box>
      <gtk::Label>{relm4::state!(counter)}</gtk::Label>
      <gtk::Button connect_clicked={on_click}>Increment</gtk::Button>
    </gtk::Box>
  )
}
``` 

Which could unpack into:

```rust
fn my_component() -> gtk::Box {
  let state = SignalCell::new(0);

  let on_click = {
    let state = state.clone();
    move |_this: &gtk::Button| {
      state.fetch_update(|current| {
        (*current) += 1;
      })
    }
  };

  let g0 = gtk::Box::builder().build();

  let g1 = gtk::Label::builder().build();
  g1.connect_unrealize({
    let subscription = state.subscribe({
      let g1 = g1.clone();
      move |value| g1.set_label(&format!("State is: {}", value));
    });
    fnonce_to_fn(move || drop(subscription));
  });
  

  let g2 = gtk::Button::builder().build();
  g2.connect_clicked(on_click);

  g0.append(&g1);
  g0.append(&g2);

  g0
}
```

And creating custom Widgets by embedding functions as partials/fragments rather than structs that extend base elements

```rust
fn app_foo() -> gtk::Button {
  relm4::rsx!(
    <gtk::Button connect_clicked={|_| println!("Hii")}>
      Hello World
    </gtk::Button>
  )
}

fn app_root() -> gtk::Box {
  relm4::rsx!(
    <gtk::Box>
      <gtk::Label>Welcome to my app</gtk::Label>
      <app_foo />
    </gtk::Box>
  )
}
```

**Comment by AaronErhardt** [comment]
I see why you would like this feature, but on the other side you need more code to achieve the same thing. Even GTK itself has a new language called [Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) which is more similar to Relm4 and compiles to the "old" XML format.

Instead of supporting this new syntax, I would rather consider adding integration for XML or Blueprint UI definitions. That said, I would be open to merge support for this anyway, but I don't have much time for working on the macros myself currently.
